Data/HashMap/Base.hs

```diff
@@ -1594,31 +1629,50 @@ mapMaybe f = mapMaybeWithKey (const f)
 -- | /O(n)/ Filter this map by retaining only elements satisfying a
 -- predicate.
 filterWithKey :: forall k v. (k -> v -> Bool) -> HashMap k v -> HashMap k v
-filterWithKey pred = filterMapAux onLeaf onColl
+filterWithKey pred = filterMapAux onLeaf
   where onLeaf t@(Leaf _ (L k v)) | pred k v = Just t
         onLeaf _ = Nothing
-
-        onColl el@(L k v) | pred k v = Just el
-        onColl _ = Nothing
 {-# INLINE filterWithKey #-}
 
 
 -- | Common implementation for 'filterWithKey' and 'mapMaybeWithKey',
 --   allowing the former to former to reuse terms.
 filterMapAux :: forall k v1 v2
               . (HashMap k v1 -> Maybe (HashMap k v2))
-             -> (Leaf k v1 -> Maybe (Leaf k v2))
              -> HashMap k v1
              -> HashMap k v2
-filterMapAux onLeaf onColl = go
+filterMapAux onLeaf = go
   where
     go Empty = Empty
     go t@Leaf{}
         | Just t' <- onLeaf t = t'
         | otherwise = Empty
     go (BitmapIndexed b ary) = filterA ary b
     go (Full ary) = filterA ary fullNodeMask
-    go (Collision h ary) = filterC ary h
+    go (Collision h l1 l2 hm) = case (onLeaf (Leaf h l1), onLeaf (Leaf h l2)) of
+      (Just (Leaf _ l1'), Just (Leaf _ l2')) -> Collision h l1' l2' $ go hm
+      (Just (Leaf _ l1'), Nothing) -> case go' hm of
+        Nothing -> Leaf h l1'
+        Just (Leaf _ l3', hm') -> Collision h l1' l3' hm'
+        _ -> error "Should not happen, can be fixed with refactoring I think."
+      (Nothing, Just (Leaf _ l2')) -> case go' hm of
+        Nothing -> Leaf h l2'
+        Just (Leaf _ l3', hm') -> Collision h l2' l3' hm'
+        _ -> error "Should not happen, can be fixed with refactoring I think."
+      (Nothing, Nothing) -> rehash h $ go hm
```

Comment:
Actually, I think I was all wrong about rehashing after all. Very sorry. Let's simplify this:

```haskell
go (Collision h l1 l2 hm) = case (onLeaf (Leaf h l1), onLeaf (Leaf h l2)) of
  (Just (Leaf _ l1'), Just (Leaf _ l2')) -> Collision h l1' l2' $ go hm
  (Just (Leaf _ lf), Nothing) -> go1 lf
  (Nothing, Just (Leaf _ lf)) -> go1 lf
  (Nothing, Nothing) -> case unconsHM (go hm) of
     UnconsEmptyHM -> Empty
     UnconsedHM l1' hm' ->
        case unconsHM hm' of
           UnconsEmptyHM -> Leaf h l1'
           UnconsedHM l2' hm'' -> Collision h l1' l2' hm''

 where
      go1 l1 = case unconsHM (go hm) of
        UnconsEmptyHM -> Leaf h l1
        UnconsedHM l2 hm' -> Collision h l1 l2 hm'
```